### PR TITLE
Refactor railsrc file location to be xdg compliant

### PR DESCRIPTION
**Pull request by quintrino**
### Summary

The XDG Base Directory Specification (which is currently used by
FOSS projects such as Git, Tmux, Pry, Rspec) provides a default
location for various file formats, including config/rc files.

This comment refactors app_generator.rb to load railsrc from
XDG_CONFIG_HOME if both XDG_CONFIG_HOME is set and rails/railrc
exists within the XDG_CONFIG_HOME location.

To maintain backwards compatibility it defaults back to ~/.railsrc
if either XDG_CONFIG_HOME is not set or there is no rails/railsrc.

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/app_generator.rb [hunk 1]
```diff
@@ -563,7 +563,15 @@ def prepare!
...
-        File.expand_path("~/.railsrc")
+        return File.expand_path("~/.railsrc") unless ENV["XDG_CONFIG_HOME"]
+
+        xdg_railsrc = ENV["XDG_CONFIG_HOME"] + "/rails/railsrc"
+        rc_file = if File.exist?(xdg_railsrc)
+          xdg_railsrc
+        else
+          "~/.railsrc"
+        end
+        File.expand_path(rc_file)
```
Comment (jonathanhefner) [inline railties/lib/rails/generators/rails/app/app_generator.rb:574]: I feel like this control flow is a little hard to follow.  What do you think about this arrangment:

```suggestion
        xdg_railsrc = File.expand_path("rails/railsrc", ENV["XDG_CONFIG_HOME"]) if ENV["XDG_CONFIG_HOME"]

        if xdg_railsrc && File.exist?(xdg_railsrc)
          xdg_railsrc
        else
          File.expand_path("~/.railsrc")
        end
```

**Review by jonathanhefner — COMMENTED** [review]
Cool!  What about a test for this?  And we should probably add mention to [USAGE](https://github.com/rails/rails/blob/42914820d91ded5b1fb285448bbde8620e00f33c/railties/lib/rails/generators/rails/app/USAGE#L6) too.

Also, we should probably do something with [this default](https://github.com/rails/rails/blob/42914820d91ded5b1fb285448bbde8620e00f33c/railties/lib/rails/commands/plugin/plugin_command.rb#L16) in `PluginCommand`, though I'm not sure of the best course of action there.

File: railties/lib/rails/generators/rails/app/app_generator.rb (hunk 1 above)
Comment (quintrino) [reply railties/lib/rails/generators/rails/app/app_generator.rb]: @jonathanhefner Thanks!

Yep, I like the the flow and readability of this better, I've updated the commits now.  

I've updated the USAGE file as well. 

Hmm, I did a search for ~/.railsrc so I missed the File.join("~", ".railsrc"), good catch!

One option might be to extract the xdg detection logic into a shared parent module perhaps? 

We could do something like Rails.rc_file but I'm guessing the Rails module itself is fairly slow to change?

Once we decide on the location for storing the logic I'll attempt to write a test for it. 

Thanks for getting back to me. This is actually my first Open Source PR. 🙂

**Comment by jonathanhefner** [comment]
@quintrino I haven't forgotten about this!

> One option might be to extract the xdg detection logic into a shared parent module perhaps?

I think this is the correct answer, but, after some thinking, I want to factor out more than just the `default_rc_file` method.  For now, though, let's go with what we have.  This PR will add XDG support to `rails new`, and I'll work on a follow-up PR that adds XDG support to `rails plugin new`.

> Once we decide on the location for storing the logic I'll attempt to write a test for it.

:+1: For now, let's test `ARGVScrubber.default_rc_file`.

Oh, also, could you [add a changelog entry](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#updating-the-changelog)?

> This is actually my first Open Source PR.

That's awesome!!  I hope for many more! :partying_face: :champagne:

File: railties/lib/rails/generators/rails/app/app_generator.rb [hunk 2]
```diff
@@ -563,7 +563,12 @@ def prepare!
       end
 
       def self.default_rc_file
-        File.expand_path("~/.railsrc")
+        xdg_railsrc = File.expand_path("rails/railsrc", ENV["XDG_CONFIG_HOME"]) if ENV["XDG_CONFIG_HOME"]
+        if xdg_railsrc && File.exist?(xdg_railsrc)
```
Comment (sinsoku) [inline railties/lib/rails/generators/rails/app/app_generator.rb:567]: According to the XDG Base Directory Specification, if `$XDG_CONFIG_HOME` is either not set or empty, you should use `$HOME/.config`.

> $XDG_CONFIG_HOME defines the base directory relative to which user specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.
>
> ref: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

It would probably work with the following code.

```ruby
xdg_config_home = ENV["XDG_CONFIG_HOME"].presence || '~/.config'
xdg_railsrc = File.expand_path("rails/railsrc", xdg_config_home)
if File.exist?(xdg_railsrc)
  xdg_railsrc
else
  File.expand_path("~/.railsrc")
end
```

Comment (jonathanhefner) [reply railties/lib/rails/generators/rails/app/app_generator.rb:567]: @sinsoku I think you are correct in terms of the spec.  The downside of falling back to `~/.config` is that it makes this feature appear to be Linux-specific (even though it isn't), which makes it less likely to be merged.

On the other hand, both [Ruby](https://github.com/ruby/ruby/blob/b7e1de117e20911ed86b229a4c2f912e469c6472/lib/rubygems/defaults.rb#L128) and [git](https://git-scm.com/docs/git-config#Documentation/git-config.txt-XDGCONFIGHOMEgitconfig) fall back to `~/.config` regardless of OS, so I think there's an argument to be made for doing the same.

**Comment by jonathanhefner** [comment]
@quintrino  Are you still interested in working on this?

**Comment by quintrino** [comment]
Apologies on the delay on this, I was made redundant from my job 2 days after I committed this so this fell off my radar. 

@sinsoku I agree with the inclusion of ~/.config as a fallback, I've added in your changes now. 

@jonathanhefner I've created a test for ARGVScrubber.default_rc_file but at the moment it's failing locally for me. 
I've been using 
`ruby -I test railties/test/generators/argv_scrubber_test.rb` to run the test file. 

I suspect it could be because of something in $LOAD_PATH or possibly because the command I'm using isn't loading all of rails repo, and defaulting to the installed gems rather than my modified code. Any thoughts on what could be causing it to load from the gem rather than the app_generator.rb file? 

Happy to hear thoughts on how to improve the testing (I've listed some of the issues I ran into with testing the existence of an exact file in the last commit to explain why I wasn't able to use tempfile (as far as I can see).

**Comment by sinsoku** [comment]
@quintrino You need to run the test in the `railties` directory.
I was able to run the test on my local machine, so I think it will probably work in your machine as well.

```console
$ cd railties
$ ruby -I test test/generators/argv_scrubber_test.rb
Run options: --seed 19374

# Running:

..........

Finished in 0.004136s, 2417.7950 runs/s, 5077.3695 assertions/s.
10 runs, 21 assertions, 0 failures, 0 errors, 0 skips
```

The test code is successful, but there are RuboCop errors.

You have to check the result of executing "RuboCop / build (pull_request)" at the bottom of this pull request and correct the code style.

File: railties/test/generators/argv_scrubber_test.rb [hunk 3]
```diff
@@ -56,6 +57,20 @@ def self.default_rc_file
...
+        ENV["XDG_CONFIG_HOME"] = xdg
+        rails_xdg_path = File.join(xdg, "rails")
+        Dir.mkdir(rails_xdg_path, 0700) unless Dir.exist?(rails_xdg_path)
+        Dir.new(rails_xdg_path)
+        railrc_path = File.join(rails_xdg_path, "railsrc")
+        railsrc = File.new(railrc_path, "w")
+        railsrc.close
+
+        scrubber = Class.new(ARGVScrubber)
+        assert_equal scrubber.default_rc_file, railrc_path
```
Comment (sinsoku) [inline railties/test/generators/argv_scrubber_test.rb:71]: It is better to clean the file as it still exists after running the test.

```ruby
  assert_equal scrubber.default_rc_file, railrc_path
ensure
  File.delete(railrc_path) if File.exist?(railrc_path)
end
```

Comment (quintrino) [reply railties/test/generators/argv_scrubber_test.rb:71]: I've added `File.delete` now.

File: railties/test/generators/argv_scrubber_test.rb [hunk 4]
```diff
@@ -56,6 +57,21 @@ def self.default_rc_file
...
+        File.delete(railrc_path) if File.exist?(railrc_path)
```
Comment (rafaelfranca) [inline railties/test/generators/argv_scrubber_test.rb:72]: Let's move to an `ensure` block and also unset the `ENV["XDG_CONFIG_HOME"]` value.

Comment (jonathanhefner) [inline railties/test/generators/argv_scrubber_test.rb:72]: @quintrino I hope you don't mind me piling on! :pray:  I felt like this test code could be a bit more focused:

```suggestion
      def test_default_rc_file_with_xdg_config_home
        Dir.mktmpdir do |dir|
          rc_file = File.join(dir, "rails/railsrc")
          FileUtils.mkdir_p(File.dirname(rc_file))
          FileUtils.touch(rc_file)

          switch_env("XDG_CONFIG_HOME", dir) do
            assert_equal rc_file, ARGVScrubber.default_rc_file
          end
        end
```

That will take care of cleaning up the file and the environment variable.  Though to use `switch_env` you'll need to add `require "env_helpers"` to the top of the file, and `include EnvHelpers` to the top of the test class ([similar to this](https://github.com/rails/rails/blob/1fd17336544a04119316c0a9dfc2b1464b4b3f30/railties/test/generators/actions_test.rb#L5-L9)).

Comment (quintrino) [reply railties/test/generators/argv_scrubber_test.rb]: @jonathanhefner I've applied these changes now, let me know if you think it needs anything else.

Comment (quintrino) [reply railties/test/generators/argv_scrubber_test.rb]: Thanks @rafaelfranca, I actually ended up switching to Jonathan Hefners code below, but good point on unsetting the ENV, I hadn't considered that.

**Review by jonathanhefner — COMMENTED** [review]
A couple of nitpicks.  Would you mind also [squashing the commits](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits)?

File: railties/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
+*   Make railsrc file location xdg-specification compliant
+
+    Rails new will load the railsrc file from XDG_CONFIG_HOME/rails/railsrc
+    It will then fallback to .config if XDG_CONFIG_HOME not specified
+    It will then fallback to HOME/.railsrc
```
[the remark this answers is not in the document]
Comment (quintrino) [reply railties/CHANGELOG.md]: Added the blank line and backticks now, if everything looks good to you then I'll squash and force-push up everything in a single commit for review.

File: railties/CHANGELOG.md [hunk 6]
```diff
@@ -1,8 +1,8 @@
 *   Make railsrc file location xdg-specification compliant
 
-    Rails new will load the railsrc file from XDG_CONFIG_HOME/rails/railsrc
-    It will then fallback to .config if XDG_CONFIG_HOME not specified
-    It will then fallback to HOME/.railsrc
+    Rails new will load the `railsrc` file from `XDG_CONFIG_HOME/rails/railsrc`.
+    It will then fallback to `.config` if `XDG_CONFIG_HOME` not specified.
+    It will then fallback to `HOME/.railsrc`.
```
Comment (jonathanhefner) [inline railties/CHANGELOG.md:5]: "Rails new" could use some backticks too!  Also, to clarify what `.config` means, what do you think about:

```suggestion
    `rails new` will now look for the default `railsrc` file at
    `$XDG_CONFIG_HOME/rails/railsrc` (or `~/.config/rails/railsrc` if
    `XDG_CONFIG_HOME` is not set).  If this file does not exist, `rails new`
    will fall back to `~/.railsrc`.
```

Comment (quintrino) [reply railties/CHANGELOG.md]: Sure, I've made these changes now.
